### PR TITLE
Fix slowness in project related OPA queries

### PR DESCRIPTION
**Pull request by tylercloke**
### :nut_and_bolt: Description: What code changed, and why?

Our project related OPA queries were very slow for ProjectsAuthorized and V2FilterAuthorizedPairs -- especially for large sets of subjects.

We made a few improvements:

1) Optimize the ProjectsAuthorized query by using lower level Rego APIs to run the partial evaluation.
2) Use set difference instead of negation in the introspect query to avoid a cross-product in `pairs_match`.
3) Memoize and prevent backtracking for `has_member`.

Both of the queries were taking so long they were timing out in the UI. Now everything should take < .5 seconds for up to 1000 subjects (probably unwise to go above that still though).

### :athletic_shoe: How to Build and Test the Change

```
chef-automate iam upgrade-to-v2 --beta2.1
export ADMINTOK=`chef-automate iam token create admin --admin`
```

Create a user with many teams. Here's a script I used you can modify:

```
cat tyler-local/add-user-to-teams.sh
#!/bin/bash
for i in {1..100}
do
   echo "Creating team $i"
   curl -H "api-token: $ADMINTOK" -XPOST -d '{"id":"newteam'$i'", "name":"newteam'$i'"}' https://a2-dev.test/apis/iam/v2beta/teams --insecure
   curl -H "api-token: $ADMINTOK" -XPOST -d '{"user_ids": ["e7b200d5-1df8-45f1-afac-fec1909dbe06"]}' "https://a2-dev.test/apis/iam/v2beta/teams/newteam$i/users:add" --insecure
done
```

If you log in as that user, you'll see many requests time out in the network tab and general brokeness.

```
rebuild components/authz-service
```

Things should be nice and snappy again for up to 1000 users.

### :white_check_mark: Checklist

- [x] Tests added/updated?
- [ ] Docs added/updated?
 
### Special Thanks

@tsandall and the OPA team!

### CODE REVIEW

File: components/authz-service/engine/opa/policy/introspection_v2.rego [hunk 1]
```diff
@@ -26,24 +26,30 @@ pair_matches_action[[pol_id, statement_id, pair]] {
 	authz.action_matches(pair.action, role_action)
 }
 
+# This causes the has_member set to be generated and memoized once instead of
+# for each policy. Moreover it prevents the backtracking from generating duplicate
+# answers which be evaluated by the resource and action matchers.
+has_member = authz.has_member
```
Comment (blakestier) [inline components/authz-service/engine/opa/policy/introspection_v2.rego]: :awesome:

File: components/authz-service/engine/opa/opa.go [hunk 2]
```diff
@@ -176,23 +176,78 @@ func (s *State) initPartialResultV2(ctx context.Context) error {
...
+	// support queries that ask for a single decision, i.e., they cannot contain
+	// variables or multiple expressions.
+	pq, err := r.Partial(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Partial evaluation returns a set of queries and support modules that have
+	// to be compiled. Add the support modules to compiler.
+	for i, module := range pq.Support {
```
[the remark this answers is not in the document]
Comment (tylercloke) [reply components/authz-service/engine/opa/opa.go]: This is all I have seen so far:

https://blog.openpolicyagent.org/partial-evaluation-162750eaf422

[the remark this answers is not in the document]
Comment (tylercloke) [reply components/authz-service/engine/opa/opa.go]: This is all I have seen so far:

https://blog.openpolicyagent.org/partial-evaluation-162750eaf422

File: components/authz-service/engine/opa/policy/introspection_v2.rego [hunk 3]
```diff
@@ -26,24 +26,30 @@ pair_matches_action[[pol_id, statement_id, pair]] {
 	authz.action_matches(pair.action, role_action)
 }
 
+# This causes the has_member set to be generated and memoized once instead of
+# for each policy. Moreover it prevents the backtracking from generating duplicate
+# answers which be evaluated by the resource and action matchers.
```
Comment (msorens) [inline components/authz-service/engine/opa/policy/introspection_v2.rego]: Not clear on the meaning due to the grammar glitch "which be"; not sure what it should read.

Comment (tsandall) [reply components/authz-service/engine/opa/policy/introspection_v2.rego]: Whoops, `s/which be/that have to be/`

**Pull request merged**